Weld Spring Boot Starter

classicPintus opened the pull request:
Hello guys,
I've created the weld-spring-boot-starter.

I've tried to follow all the guidelines. Hope to be useful, and sorry if there're mistakes.

**ACTIONS TO USE IT**

1. Add the weld-spring-boot-starter dependency
```
<dependency>
    <groupId>org.joinfaces</groupId>
    <artifactId>weld-spring-boot-starter</artifactId>
</dependency>
```

2. Override the omnifaces.version to a version 2.x or 3.x
`<omnifaces.version>3.0</omnifaces.version>`

3. Start the project

**OPEN POINT**
By checking [this answer](https://stackoverflow.com/questions/48420941/omnifaces-2-x-fails-when-used-in-a-spring-boot-environment-with-weld/48474573#48474573) I've found that Omnifaces doesn't log that is started.
After checking seems that this code is needed

```
public class OmnifacesServletContextInitializer implements ServletContextInitializer
{
    private ServletContainerInitializer servletContainerInitializer;

    @Override
    public void onStartup(ServletContext sc) throws ServletException
    {
        ServletContainerInitializer servletContainerInitializer = getServletContainerInitializer();
        servletContainerInitializer.onStartup(null, sc);
    }

    private ServletContainerInitializer getServletContainerInitializer()
    {
        if (servletContainerInitializer == null)
            servletContainerInitializer = new ApplicationInitializer();

        return servletContainerInitializer;
    }
}
```

```
@Bean
public ServletContextInitializer omnifacesServletContextInitializer()
{
    return new OmnifacesServletContextInitializer();
}
```

At the moment I don't know if this piece of code is needed or not. @BalusC can answer us I think.
If this code is needed, just keep in mind that the ApplicationInitializer class is not there in the 1.x versions of Omnifaces.
IMHO Joinfaces can forget the 1.x and focus on the 3.x. But it's only my opinion.

closes #187

larsgrefer commented on joinfaces-autoconfigure/src/main/java/org/joinfaces/autoconfigure/weld/WeldSpringBootAutoConfiguration.java:
```diff
@@ -0,0 +1,41 @@
+
+@Configuration
+@ConditionalOnWebApplication
+@ConditionalOnClass(name = "org.jboss.weld.environment.servlet.EnhancedListener")
```
You can directly reference the Class here.

classicPintus replied:
Done

larsgrefer commented on joinfaces-autoconfigure/src/main/java/org/joinfaces/autoconfigure/weld/package-info.java:
```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright 2016-2016 the original author or authors.
```
Please adjust the year here.

classicPintus replied:
Like this Lars?
Copyright 2016-2018 the original author or authors.

classicPintus replied:
Done

classicPintus commented:
I've made the modifications asked by @larsgrefer

persapiens commented:
I'm not with my desktop this week. I will check your PR as soon as possible.

larsgrefer commented on joinfaces-dependencies/pom.xml:
```diff
@@ -1,602 +1,627 @@
-        <version>2.0.0.RC1</version>
-        <relativePath /> <!-- lookup parent from repository -->
-    </parent>
+	<modelVersion>4.0.0</modelVersion>
```
The indention here should be kept 4 spaces instead of tabs

larsgrefer commented on joinfaces-starters/weld-spring-boot-starter/pom.xml:
```diff
@@ -0,0 +1,24 @@
+        <artifactId>joinfaces-starters</artifactId>
+        <version>3.0.0.RC2-SNAPSHOT</version>
+    </parent>
+    <artifactId>weld-spring-boot-starter</artifactId>
```
Aren't `weld-spring-boot-starter` and `omnifaces-spring-boot-starter` mutually dependent now?

classicPintus replied:
Hi Lars,
as per my understanding, joinfaces declares to support omnifaces 1.x. With this omnifaces version, Weld is not needed.

Futhermore, weld can be used for everything that needs CDI, and Omnifaces is not the only library that needs CDI. Potentially you can decide to use weld in order to make working libraries different than omnifaces

larsgrefer replied:
Then we should have 2 omifaces starters, one for omnifaces 1.x and another one for the current omnifaces version.

classicPintus replied:
Are we sure? I'm asking for some reasons.

1. Currently switch from omnifaces 1 to omnifaces 3 is very easy. We only need to override the omnifaces.version property and add the dependency to the weld-spring-boot-starter.

2. Directly from the omnifaces site "For users on yet more outdated environments who can’t/won’t use CDI, use 1.14.1 instead. It doesn’t contain anything from CDI nor new things which were added in 2.x, but it does contain enhancements and bugfixes to existing 1.x things. Note that there is no 1.15 nor 1.16. The 1.14.1 is latest version of the 1.x branch which is also in maintenance mode." Seems that Omnifaces 1.x is abandoned. So if joinfaces decides to abandon Omnifaces as well, the second starter is not needed. 

So basically, I think I'm able to do that. the second starter. I'm only asking in order to be sure that is the direction that Joinfaces needs.

larsgrefer commented on joinfaces-dependencies/pom.xml:
```diff
@@ -348,6 +348,25 @@
                 <artifactId>richfaces-spring-boot-starter</artifactId>
                 <version>3.0.0.RC2-SNAPSHOT</version>
+            </dependency>
+			<dependency>
```
This line still contains tabs

larsgrefer commented on joinfaces-dependencies/pom.xml:
```diff
@@ -395,7 +414,7 @@
                     <configuration>
                         <targetJdk>${java.source}</targetJdk>
-                        <includeTests>true</includeTests>
+						<includeTests>true</includeTests>
```
This line still contains tabs

larsgrefer commented on joinfaces-dependencies/pom.xml:
```diff
@@ -246,7 +247,7 @@
             </dependency>
             <dependency>
-                <groupId>de.larmic.butterfaces</groupId>
+				<groupId>de.larmic.butterfaces</groupId>
```
This line still contains tabs

larsgrefer commented on joinfaces-dependencies/pom.xml:
```diff
@@ -34,11 +34,12 @@
         <richfaces.version>4.5.17.Final</richfaces.version>
 
         <cdi.version>2.0</cdi.version>
+		<weld.version>3.0.3.Final</weld.version>
```
This line still contains tabs

larsgrefer commented on joinfaces-dependencies/pom.xml:
```diff
@@ -34,11 +34,12 @@
         <angularfaces.version>2.1.12</angularfaces.version>
 
-        <butterfaces.version>2.1.25</butterfaces.version>
+		<butterfaces.version>2.1.25</butterfaces.version>
```
This line still contains tabs

larsgrefer commented:
Before we can merge this, you have to merge our current master branch into your master branch

larsgrefer commented:
@persapiens What do you think? Should we support both omnifaces 1.x and 3.x? Should we drop 1.x in favor of 3.x?

From Joinfaces' point of view 1.x and 3.x are not binary compatible. For 1.x Joinfaces currently registers the `org.omnifaces.facesviews.FacesViewsInitializer` for embedded servlet-containers. In omnifaces 3.x this class doesn't exist, but 3.x has `org.omnifaces.ApplicationInitializer` which we would had to handle instead. So Joinfaces would have to be compiled against omnifaces 1.x and 3.x at the same time. Im not sure if maven allows this (but we could use two autoconfigure modules or put some autoconfiguration directly into the starters).

When dropping 1.x in favor of 3.x, `omnifaces-spring-boot-starter` should depend on `weld-spring-boot-starter`. And `jsf-spring-boot-starter` should *not* depend on `omnifaces-spring-boot-starter` anymore, because not everyone using Joinfaces wants weld too.

persapiens commented:
hi @classicPintus @larsgrefer 

JoinFaces was initially designed to solve JSF and Spring Boot integration.
I imagine that fully CDI and Spring Boot integration is another great challenge and it could be solved by another project, probably JoinBeans ;) 

Anyway, we could start CDI and Spring Boot integration here at JoinFaces and we spin off later.

`Should we support both OmniFaces 1.x and 3.x? `

I think it would be the ideal solution because CDI implementation is not (yet) required by JSF (2.x). More, I think 1.x should be default because it is better tested. When 3.x becomes stable enough, we can switch the default one.

`From Joinfaces' point of view 1.x and 3.x are not binary compatible.`

If maven does not support compile against two versions of the same library, let's create another autoconfigure module specific to omnifaces 1.x instead of put autoconfiguration directly into the starters.

I propose the following changes:
- move current omnifaces 1.x autoconfiguration from joinfaces-autoconfigure to joinfaces-omnifaces1-autoconfigure module 
- rename current omnifaces-spring-boot-starter to omnifaces1-spring-boot-starter. jsf-spring-boot-starter depends on omnifaces1-spring-boot-starter
- include omnifaces 3.x and weld autoconfiguration classes into joinfaces-autoconfigure module. Of course, using classes provided by this PR.
- create weld-spring-boot-starter
- create new omnifaces-spring-boot-starter using omnifaces 3.x. It depends on weld-spring-boot-starter

So, jsf-spring-boot-starter depends on omnifaces1-spring-boot-starter. When someone wants to use omnifaces 3.x, it has to remove omnifaces1-spring-boot-starter dependency and add omnifaces-spring-boot-starter.

What do you think?

classicPintus commented:
Hi @persapiens 
I agree with you, I've only some doubts about the priorities. I'll try to explain my mindset.

By using "my" fork, anyone can have the freedom to choose everything he wants (I've tried to explain that in the first post of the current pull request)
- JoinFaces + Omnifaces 1.x without / with weld
- JoinFaces + Omnifaces 2.x with / without weld
- JoinFaces + Omnifaces 3.x with Weld

So basically all the modifications proposed by you in the previous post, makes a lot of sense but from the final user point of view, doesn't adds any functionality.

In my opinion should be better to focus on something else.
For example I'm not switching my projects to joinfaces because seems that (I've read something [here](https://github.com/joinfaces/joinfaces/issues/356) )  the facesServlet mapping can't be customized.
Maybe should be a good idea to implement this functionality?

If you disagree no problem. 
I'll try to follow your tips in order to create the best I can

larsgrefer commented:
@classicPintus How is it possible to use Omnifaces 2.x without weld? According to the Omnifaces website, CDI is required for both 2.x and 3.x.

Also Omnifaces 2.x and 3.x won't work on embedded containers (jar-packaging) without additional auto-configuration which is currently not included in joinfaces.

classicPintus commented:
@larsgrefer 
Yes, confirm. I was misleading the optional depedency with weld of jsf 2.2 with the same version of omnifaces.

About the second point, maybe I'm missing something. Currently I'm able to use omnifaces 3.0 with joinfaces. I've not tried all the omnifaces tags, but for example <o:form> works.
Your pull request #413 contains basically my open point that you can see on the top. 
But if now we're sure that code is needed, so we've an open point less.

persapiens commented:
hi @classicPintus 

`So basically all the modifications proposed by you in the previous post, makes a lot of sense but from the final user point of view, doesn't add any functionality.`

I disagree. In my opinion, full CDI and Spring Boot integration is a huge feature. You are solving one of the [oldest joinfaces issue](https://github.com/joinfaces/joinfaces/issues/187). One day, it will be required. So, this PR will probably guarantee JoinFaces evolution because we this feature will be mandatory.

More, full CDI and Spring Boot integration open new possibilities beyond joinfaces scope. For example, using many CDI libraries would be used together with Spring.

`For example, I'm not switching my projects to joinfaces because seems that (I've read something here ) the facesServlet mapping can't be customized. Maybe should be a good idea to implement this functionality?`

JoinFaces is opensource. We work here as we can. Of course, your feedback and effort influence our [Roadmap](https://github.com/joinfaces/joinfaces/wiki/Roadmap). This PR got attention because you submit it ;)

What do you think about finishing this PR and after working with #356 ?

classicPintus commented:
Great debate here. Very useful.
So if you agree I'll do the following steps.

- move current omnifaces 1.x autoconfiguration from joinfaces-autoconfigure to joinfaces-omnifaces1-autoconfigure module

- rename current omnifaces-spring-boot-starter to omnifaces1-spring-boot-starter. jsf-spring-boot-starter depends on omnifaces1-spring-boot-starter

- include omnifaces 3.x and weld autoconfiguration classes into joinfaces-autoconfigure module. Of course, using classes provided by this PR.

- create weld-spring-boot-starter

- create new omnifaces-spring-boot-starter using omnifaces 3.x. It depends on weld-spring-boot-starter

Once I've your last confirmation I'll start.

persapiens commented:
Go ahead @classicPintus : may the force be with you.

Note that @larsgrefer has submitted #413 . It may be helpful to all of us.

larsgrefer commented:
> - move current omnifaces 1.x autoconfiguration from joinfaces-autoconfigure to joinfaces-omnifaces1-autoconfigure module

This should not be necessary with #413

classicPintus commented:
Is it possible to include in this PR, also the #413 in order to avoid one more merge?
Sorry but for me are the first experiences with github

larsgrefer commented:
@classicPintus Sure:

`git pull https://github.com/larsgrefer/joinfaces.git feature/omnifacesAutoconfiguration`

persapiens commented:
hi @classicPintus ,

    Could you fix conflicting files compared to master branch? I tried here, but tests are broken after trying some fixes.

classicPintus commented:
I'm checking now

classicPintus commented:
The conflict in spring.factories is because I've added the following line
`\,org.joinfaces.autoconfigure.weld.WeldSpringBootAutoConfiguration`

In the joinfaces-dependencies/pom.xml I've added the following lines

```
<weld.version>3.0.3.Final</weld.version>

<dependency>
                <groupId>org.jboss.weld.servlet</groupId>
                    <artifactId>weld-servlet-core</artifactId>
                    <version>${weld.version}</version>
                    <exclusions>
                        <exclusion>
                            <groupId>org.jboss.spec.javax.el</groupId>
                            <artifactId>jboss-el-api_3.0_spec</artifactId>
                        </exclusion>
                        <exclusion>
                            <groupId>org.jboss.logging</groupId>
                            <artifactId>jboss-logging</artifactId>
                        </exclusion>
                        <exclusion>
                            <groupId>org.jboss.spec.javax.interceptor</groupId>
                            <artifactId>jboss-interceptors-api_1.2_spec</artifactId>
                        </exclusion>
                    </exclusions>
            </dependency>

<dependency>
                <groupId>org.joinfaces</groupId>
                <artifactId>weld-spring-boot-starter</artifactId>
                <version>3.0.3-SNAPSHOT</version>
            </dependency>
```

and the modification that created the omnifaces1-spring-boot starter

The joinfaces-starters/omnifaces-spring-boot-starter/pom.xml doesn't exists at the moment because there's the joinfaces-starters/omnifaces1-spring-boot-starter/pom.xml

classicPintus commented:
In [my fork](https://github.com/classicPintus/joinfaces) I've added a file named weld.patch.
Maybe is useful

I've tried to simulate a merge in my local environment and all the tests are passing. 
If there're more problems we can try to be faster by using skype for example. My username on skype is gavino.pintus

larsgrefer commented:
@classicPintus I just had a look at your commit history. Something seems wrong with your merges.
The commit message says `MERGE`, but they aren't actually merge commits. (They don't have two parent commits).

The fact that these _fake_-merge commits contain the changes of another commit, but don't declare this commit as their parent, makes futher merges very complicated.

classicPintus commented:
Thank you Lars. I'll increase my skills from the git point of view.

I've saw a lot of push in the fork. I've tried the usual test with the [joinfaces example](https://github.com/joinfaces/joinfaces-maven-jar-example) but it's not working because the context is not able to start. The error is the following but I suppose that the real error is elsewhere

java.lang.IllegalStateException: Failed to load property source from location 'classpath:/application.yml'
	at org.springframework.boot.context.config.ConfigFileApplicationListener$Loader.load(ConfigFileApplicationListener.java:535)
	at org.springframework.boot.context.config.ConfigFileApplicationListener$Loader.loadForFileExtension(ConfigFileApplicationListener.java:494)
	at org.springframework.boot.context.config.ConfigFileApplicationListener$Loader.load(ConfigFileApplicationListener.java:462)
	at org.springframework.boot.context.config.ConfigFileApplicationListener$Loader.lambda$null$4(ConfigFileApplicationListener.java:444)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at org.springframework.boot.context.config.ConfigFileApplicationListener$Loader.lambda$load$5(ConfigFileApplicationListener.java:443)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at org.springframework.boot.context.config.ConfigFileApplicationListener$Loader.load(ConfigFileApplicationListener.java:440)
	at org.springframework.boot.context.config.ConfigFileApplicationListener$Loader.load(ConfigFileApplicationListener.java:331)
	at org.springframework.boot.context.config.ConfigFileApplicationListener.addPropertySources(ConfigFileApplicationListener.java:213)
	at org.springframework.boot.context.config.ConfigFileApplicationListener.postProcessEnvironment(ConfigFileApplicationListener.java:196)
	at org.springframework.boot.context.config.ConfigFileApplicationListener.onApplicationEnvironmentPreparedEvent(ConfigFileApplicationListener.java:183)
	at org.springframework.boot.context.config.ConfigFileApplicationListener.onApplicationEvent(ConfigFileApplicationListener.java:169)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:172)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:165)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:139)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:127)
	at org.springframework.boot.context.event.EventPublishingRunListener.environmentPrepared(EventPublishingRunListener.java:74)
	at org.springframework.boot.SpringApplicationRunListeners.environmentPrepared(SpringApplicationRunListeners.java:54)
	at org.springframework.boot.SpringApplication.prepareEnvironment(SpringApplication.java:351)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:317)
	at org.springframework.boot.test.context.SpringBootContextLoader.loadContext(SpringBootContextLoader.java:138)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContextInternal(DefaultCacheAwareContextLoaderDelegate.java:99)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:117)
	at org.springframework.test.context.support.DefaultTestContext.getApplicationContext(DefaultTestContext.java:107)
	at org.springframework.boot.test.autoconfigure.SpringBootDependencyInjectionTestExecutionListener.outputConditionEvaluationReport(SpringBootDependencyInjectionTestExecutionListener.java:54)
	at org.springframework.boot.test.autoconfigure.SpringBootDependencyInjectionTestExecutionListener.prepareTestInstance(SpringBootDependencyInjectionTestExecutionListener.java:47)
	at org.springframework.test.context.TestContextManager.prepareTestInstance(TestContextManager.java:242)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.createTest(SpringJUnit4ClassRunner.java:227)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner$1.runReflectiveCall(SpringJUnit4ClassRunner.java:289)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.methodBlock(SpringJUnit4ClassRunner.java:291)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:246)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:97)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:190)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:369)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:275)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:239)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:160)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:373)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:334)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:119)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:407)
Caused by: org.yaml.snakeyaml.scanner.ScannerException: while scanning for the next token
found character '@' that cannot start any token. (Do not use @ for indentation)
 in 'reader', line 23, column 12:
      version: @omnifaces.version@
               ^

	at org.yaml.snakeyaml.scanner.ScannerImpl.fetchMoreTokens(ScannerImpl.java:421)
	at org.yaml.snakeyaml.scanner.ScannerImpl.checkToken(ScannerImpl.java:226)
	at org.yaml.snakeyaml.parser.ParserImpl$ParseBlockMappingValue.produce(ParserImpl.java:585)
	at org.yaml.snakeyaml.parser.ParserImpl.peekEvent(ParserImpl.java:157)
	at org.yaml.snakeyaml.parser.ParserImpl.checkEvent(ParserImpl.java:147)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:133)
	at org.yaml.snakeyaml.composer.Composer.composeValueNode(Composer.java:249)
	at org.yaml.snakeyaml.composer.Composer.composeMappingChildren(Composer.java:240)
	at org.yaml.snakeyaml.composer.Composer.composeMappingNode(Composer.java:228)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:154)
	at org.yaml.snakeyaml.composer.Composer.composeValueNode(Composer.java:249)
	at org.yaml.snakeyaml.composer.Composer.composeMappingChildren(Composer.java:240)
	at org.yaml.snakeyaml.composer.Composer.composeMappingNode(Composer.java:228)
	at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:154)
	at org.yaml.snakeyaml.composer.Composer.composeDocument(Composer.java:122)
	at org.yaml.snakeyaml.composer.Composer.getNode(Composer.java:84)
	at org.yaml.snakeyaml.constructor.BaseConstructor.getData(BaseConstructor.java:123)
	at org.yaml.snakeyaml.Yaml$1.next(Yaml.java:547)
	at org.springframework.beans.factory.config.YamlProcessor.process(YamlProcessor.java:161)
	at org.springframework.beans.factory.config.YamlProcessor.process(YamlProcessor.java:139)
	at org.springframework.boot.env.OriginTrackedYamlLoader.load(OriginTrackedYamlLoader.java:72)
	at org.springframework.boot.env.YamlPropertySourceLoader.load(YamlPropertySourceLoader.java:50)
	at org.springframework.boot.context.config.ConfigFileApplicationListener$Loader.loadDocuments(ConfigFileApplicationListener.java:542)
	at org.springframework.boot.context.config.ConfigFileApplicationListener$Loader.load(ConfigFileApplicationListener.java:515)

Futhermore I've saw that in omnifaces 3.0 the class CacheInitializerListener and Html5RenderKit doens't exists in Omnifaces 3.0, and the class CacheInstancePerScopeProvider have changed package.
Is it a problem when using Omnifaces 3.0?

classicPintus commented:
I've performed some investigations.

The problem is that the joinfaces example required the maven property <omnifaces.version> that was not there.
I've added the omnifaces.version property for each omnifaces starter, but the property is not recognized.

I'm reading the spring boot documentation in order to understand in deep why the property is not recognized

classicPintus commented:
After reading tons of maven documentation, seems that this behaviour is by design.
The inheritance is always from parent to child. Not the opposite.

For the [example](https://github.com/joinfaces/joinfaces-maven-jar-example) the parent is joinfaces-parent (and then joinfaces-dependencies also) in which are declared some maven properties. The property omnifaces.version is not there.

If I remember correctly, before there was the omnifaces.version property with the value 1.14.1

larsgrefer commented:
I've removed the `omnifaces.version` property (was `1.14.1` before).

Since we are supporting omnifaces 1.x and 3.x from now on, there is no "obvious" right value for this property.

classicPintus commented:
Yes, I agree.

Just for information, now in order to make working the example we need to declare the omnifaces.version property in the pom of the example, and also the omnifaces starter needed.

I was checking if all the steps are completed. In the following point, the bold section is not completed

- rename current omnifaces-spring-boot-starter to omnifaces1-spring-boot-starter. **jsf-spring-boot-starter depends on omnifaces1-spring-boot-starter**

Is it correct?

persapiens commented:
Great job here @classicPintus and @larsgrefer.

Some minor suggestions:

First, what do you think about creating two properties (omnifaces1.version and omnifaces3.version) at joinfaces-dependencies module and using them at joinfaces-autoconfigure, omnifaces1-spring-boot-starter and omnifaces3-spring-boot-starter? Current implementation sets 1.14.1 version twice at joinfaces-autoconfigure and omnifaces1-spring-starter modules. It is error-prone.

- joinfaces-dependencies/pom.xml

```
    <omnifaces1.version>1.14.1<omnifaces1.version>
    <omnifaces3.version>3.0</omnifaces3.version>
```

- joinfaces-autoconfigure/pom.xml

```
    <dependency>
         <groupId>org.omnifaces</groupId>
         <artifactId>omnifaces</artifactId>
         <version>${omnifaces1.version}</version>
         <optional>true</optional>
    </dependency>
```

- joinfaces-starters/omnifaces1-spring-boot-starter/pom.xml

```
    <omnifaces.version>${omnifaces1.version}<omnifaces.version>
```

- joinfaces-starters/omnifaces3-spring-boot-starter/pom.xml

```
    <omnifaces.version>${omnifaces3.version}<omnifaces.version>
```

Second, In my opinion, jsf-spring-boot-starter is meta starter using default 'starters'. It should choose default omnifaces starter too. I vote `omnifaces1-spring-boot-starter` because it will keep current joinfaces behavior.

larsgrefer commented:
Using two different properties is a good idea 👍

I'm not sure if `jsf-spring-boot-starter` should pull omnifaces at all. If I include `jsf-spring-boot-starter` in my project I expect to get plain JSF (and I don't expect to get omnifaces, too). But keeping the current behaviour is a good point, too.

persapiens commented:
Omnifaces declares itself as

> OmniFaces is a utility library for JSF 2 that focusses on utilities that ease everyday tasks with the standard JSF API.

It works with all jsf component libraries. And it is frequently in my apps. But adding omnifaces starter is not a problem.

Anyway @classicPintus , your vote decides ;)

classicPintus commented:
In fact doesn't change so much for us and also for the joinfaces users. In my opinion we can mantain the current solution.
If in the future we dislike that for some reason we can change whenever we want.

persapiens commented:
So, let's resume final tasks before merging:

1. Create omnifaces1.version and omnifaces3.version according to my previous comment

1. jsf-spring-boot-starter should pull omnifaces1-spring-boot-starter

1. fix conflict with latest pushed code.

We will merge as soon as you complete it.

larsgrefer commented:
@persapiens The final tasks are completed

persapiens commented:
Fantastic work here @classicPintus and @larsgrefer 

This joinfaces issue was the most discussed one.

I am proud of what we have done together.

Thank you very much.

classicPintus commented:
Hope to be useful also for the future

Tosken1337 commented:
Hi guys,

thanks for this great work.
Using the omnifaces3 starter together with weld now brings a full CDI with spring boot?
Are there any issues using it together with spring or anything we have to take care of?

Thanks
